File: src/map/area.rs
use crate::{
    files::files::create_dir,
    shared::{error::BaseError, json::JSON},
};
use serde::{Deserialize, Serialize};
use std::path::Path;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Area {
    pub name: String,
    childrens: Vec<Area>,
}

impl JSON for Area {}

impl Area {
    pub fn new(name: &str) -> Self {
        Area {
            name: name.to_string(),
            childrens: Vec::new(),
        }
    }

    pub fn add_child(&mut self, child: Area) {
        self.childrens.push(child);
    }

    pub fn add_childrens(&mut self, children: Vec<Area>) {
        self.childrens.extend(children);
    }

    pub fn init_area<T>(&self, root: T)
    where
        T: Into<String>,
    {
        self.create(root);
    }

    fn create<T>(&self, parent: T)
    where
        T: Into<String>,
    {
        let parent = parent.into();
        if let Some(p) = Path::new(parent.as_str()).join(&self.name).to_str() {
            if let Err(e) = create_dir(p) {
                eprintln!("Error creating directory {}: {}", p, e.get_message());
            } else {
                for child in &self.childrens {
                    child.create(p);
                }
            }
        }
    }
}

File: map/Area.cs
using System.Text.Json;

namespace map
{
  public class Area(string name)
  {
    protected string name = name;
    protected Area[] child = [];

    public string Name
    {
      get { return name; }
      set { name = value; }
    }

    public Area[] Child
    {
      get { return child; }
      set { child = value; }
    }

    protected void Create(string parent)
    {
      string path = Path.Join(parent, Name);
      Directory.CreateDirectory(path);

      for (int i = 0; i < child.Length; i++)
      {
        child[i].Create(path);
      }
    }

    public string ToJson()
    {
      return JsonSerializer.Serialize<Area>(this);
    }

    public static Area FromFile(string filename)
    {
      string json = File.ReadAllText(filename);

      if (!string.IsNullOrEmpty(json))
      {
        Area? area = JsonSerializer.Deserialize<Area>(json);

        if (area is not null) return area;
      }

      return new("");
    }

    public void ToFile(string? path = null)
    {
      string content = ToJson();
      string filename = Name + ".json";

      if (!string.IsNullOrEmpty(path))
      {
        filename = path;
      }
      File.WriteAllText(filename, content);
    }

    public void Push(Area[] areas)
    {
      child = child.Concat(areas).ToArray();
    }

    public void Init()
    {
      if (!string.IsNullOrEmpty(Name))
      {
        Create("");
      }
    }
  }
}
